[SPARK-37124][SQL] Support RowToColumnarExec with Arrow format

xuechendi opened the pull request:
### What changes were proposed in this pull request?
This Jira is aim to support Arrow format in RowToColumnarExec.

### Why are the changes needed?
Current ArrowColumnVector is not fully equivalent to OnHeap/OffHeapColumnVector in spark, so RowToColumnarExec doesn't support write to Arrow format so far.

since Arrow API is now being more stable, and using pandas udf will perform much better than python udf.

### What has been done in this pull request?
I am  proposing to support RowToColumnarExec with Arrow.

What I did in this PR is to add a load api in ArrowColumnVector to load arrowRecordBatch to ArrowColumnVector, then called inside RowToColumnarExec doExecute.

### How was this patch tested?
UTs are also added to test this new API and RowToColumnarExec with ArrowFormat.

### Does this PR introduce _any_ user-facing change?
NO

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi committed:
Support RowToColumnarExec to write to Arrow

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented:
@cloud-fan @viirya @sunchao , hi, all, I realized that what you guys said makes sense to me, instead of writing data to arrow like what other WritableColumnVector does, I can simply enabling a Load API in ArrowColumnVector and called from RowToColumnarExec, and other operators only needs 'get' APIs which already exsiting in ArrowColumnVector.
So I updated my PR and wrote some UT by testing load APIs and RowToColumnarExec operator.

BryanCutler commented on sql/catalyst/src/main/java/org/apache/spark/sql/vectorized/ArrowColumnVector.java:
```diff
@@ -38,6 +43,29 @@
   private ArrowColumnVector[] childColumns;
 
+  public static ArrowColumnVector[] loadColumns(StructType schema,
+                                                ArrowRecordBatch recordBatch) {
```
I'm not really sure this makes sense here, it seems like it's mainly to convert an `ArrowRecordBatch` to Spark `ColumnarBatch`. That is basically what's done in `ArrowConverters.fromBatchIterator`, maybe you could move that to it's own function there?

xuechendi replied:
@BryanCutler, yes, the codes is very similiar because that is how arrow designed to load ArrowRecordBatch to java List[Vector], do you mean I should put this loadColumns function to ArrowConverters, and named as something like <code>ArrowConverters.toColumnarBatchIterator : Iterator[ColumnarBatch]</code> ? I am OK to do that, I just felt it could fit in both scope, in <code>ArrowConverters</code> or in <code>ArrowColumnVector</code>

BryanCutler replied:
I was more suggesting something like adding `ArrowConverters.toColumnarBatch(recordBatch: ArrowRecordBatch): ColumnarBatch`. That is already basically one internally in `ArrowConverters.fromBatchIterator`, so instead of duplicating that same conversion just move the existing it to a function if it fits your needs too.

BryanCutler commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -458,6 +462,34 @@ case class RowToColumnarExec(child: SparkPlan) extends RowToColumnarTransition {
+        val context = TaskContext.get()
+        val allocator = ArrowUtils.getDefaultAllocator
+        val bytesIterator = ArrowConverters
+          .toBatchIterator(rowIterator, localSchema, maxRecordsPerBatch, timeZoneId, context)
```
I'm not quite sure of your use case for this code, could you explain a bit more what this does for you, especially why serialize the record batch to a byte array?

xuechendi replied:
@BryanCutler , my goal here is to convert Iterator[InternalRow] to Iterator[ColumnarBatch] with ArrowColumnVectors as internal ColumnVector. The reason that I converted it to byte arrays first is inspired by @sunchao in previous comment, Quoted: 
>  instead of having a writable Arrow vector, currently other projects such as Iceberg and Spark RAPIDS choose to first construct the Arrow ValueVectors and then wrap them to Spark with ArrowColumnVector."

So I figured when I need to read from internalRow and write to arrow columnar format, there will be two ways to make that happen, 1) one is to read and write to ArrowRecordBatch and load as ColumnarBatch; 2) the other option is to use Arrow JAVA put APIs to write. 
And from @sunchao , I think option 1 would make more sense since Spark already has `ArrowConverters.toBatchIterator` function

HyukjinKwon replied:
Same question here too. Why would we need another conversion step of Arrow format between `InternalRow` and `ColumnarBatch`? Are you looking for a Scala version of https://github.com/apache/spark/pull/34505?

xuechendi replied:
@HyukjinKwon , I looked into your PR and I am not sure what is the difference between your mapInArrow vs using pandas_udf, so I want to explain my PR vs pandas_udf here to avoid misunderstanding of your PR. 

Compared passing internalRows to pandas_udf, yes, the implementation here is very similiar while purpose is very different. My PR here is to support RowToColumnarExec sparkplan with arrow. By doing so, the following operators like sort, join, aggregate, project/filter can do computation based on Arrow format instead of using On/OffHeapColumnVector. 

And I think pandas_udf is used to do data transformation using pandas built-in computation or calling other thirdparty python package.

Not sure if I answered your question?

HyukjinKwon replied:
Yeah, I am working on Scala version of it at SPARK-37229, and it will require to have a Spark physical plan that converts rows (or Columnar) to Arrow, and Arrow to Columnar. I am wondering if you can leverage this physical plan via using Spark extensions.

xuechendi replied:
@HyukjinKwon , yes, That is what we do, we have some implemented operators and plug into spark using Spark Extension configuration + columnarRules to replace original row physical plans, that is why we need RowToColumnarExec and ColumnarToRowExec to do InternalRow and ArrowColumnarBatch conversion. I think one difference with mapInArrow(scala version) is we want to leverage the built-in attribute supportColumnar to avoid code changes to spark application.

HyukjinKwon replied:
Will you be able to add a separate `SparkPlan` that converts rows (or Columnar) to Arrow, and Arrow to Columnar instead of adding a configuration? Then I think I can leverage that physical plan in `mapInArrow`, and the change won't be invasive too.

For example,  you can add some physical plans like:

```scala
ColumnarToArrowExec
ArrowToColumnarExec
```

and then you use it via `SparkExtensions`. Does it make sense?

xuechendi replied:
@HyukjinKwon , `ArrowToColumnarExec`  is not needed because `ArrowColumnVector` is derived from `ColumnVector`, so current `ColumnarToRowExec` is capable of converting Arrow ColumnarBatch to InternalRow. 
And for `ColumnarToArrowExec`, there is some built-in rules [codes link](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala#L512-L558) will add a  `ColumnarToRowExec` between two SparkPlans, if we use a new SparkPlan then we need to change above codes as well. So I am hoping we don't add these two new physical plans.
For mapInArrow case, how about we add a new parameter in RowToColumnarExec to tell it to use Arrow.

xuechendi replied:
@HyukjinKwon , I just realized I misunderstand what you meant, what do you mean by 'ColumnarToArrow` and `ArrowToColumnar`? so Arrow here is referring to Array Bytes of ArrowRecordBatch? And Columnar is ArrowColumnVector?

HyukjinKwon replied:
Oh, okay. I guess there was a bit of miscommunication here.

I am more worrying about symmetry, and I am more talking about SQL physical plan level. I am thinking about having plans such as:

```scala
RowsToColumnarExec
ColumnarToRowsExec
ArrowToColumnarExec
ColumnarToArrowExec
```

What you want to do is I guess use Arrow-backed Columnar instances, right? You could, for example, pattern match on `RowsToColumnarExec(child)` and replace to something like `ColumnarToArrowExec(RowsToColumnarExec(child))` in Spark extensions.

I presume the overhead between `ColumnarBatch` and `ArrowBatch` is trivial.

HyukjinKwon replied:
or you could replace `RowsToColumnarExec(child)` to `RowsToArrowExec(child)` as an example.

xuechendi replied:
yes, my question is why we need to add two extra physical plans?
If we make ArrowColumnVector as another backend of ColumnarBatch in `RowsToColumnarExec`, it can also be used by scala version of `mapInArrow`, doesn't it?
And another thing I am confused here is what is the `Arrow ` and `Columnar`  you are referring to in `ArrowToColumnarExec` and `ColumnarToArrowExec `context? To me, arrow is columnar, so are your talking about adding two Physical Plans to convert between `ArrowRecordBatch(type of RDD[Byte[Array]])` and `ColumnarBatch(type of RDD[ColumnarBatch])`

HyukjinKwon replied:
Okay, so there are:
1. `ColumnarBatch`
2. `ColumnarBatch` backed by Arrow
3. `ArrowRecordBatch`

and what this PR does creates 2. from rows (instead of 1.). More specifically: `Rows -> 2` transition.

I was thinking something like:

`RowsToArrowExec` converts rows to 3. and `ArrowToColumnarExec` converts 2. More specifically `Rows -> 3 -> 2` transition.

I haven't taken a deeper look but I was thinking about that we should better have such physical plans and separations. I have been looking through related Arrow feature requests .. and my guts say it would be very useful to have these. At least I would like to reuse it.

Does it make sense?

xuechendi replied:
I see, I am not sure if you are familiar with [toArrowBatchRdd](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L3857-L3872) in Dataset,  it really looks like what `RowsToArrowExec` is.
And if your are looking for 

> converts 2. More specifically Rows -> 3 -> 2 transition.

I think that is what I am proposing in `RowToColumnarExec`, so if I think we can add a parameter in `RowToColumnarExec` to specify Arrow as backend, it already gets what we wanted.

After all, if you think adding `RowToArrowExec`  won't be duplicated to `Dataset.toArrowBatchRdd`, I am ok to add these two new physical plans, then do a combination of physical plan of converting Row -> 3 -> 2. I can do it in a seperate PR , but I still hope to make sure adding two new physical plans is a desired way. @cloud-fan @sunchao @BryanCutler

HyukjinKwon replied:
(FWIW, I think we can deduplicate `toArrowBatchRdd` by using that new physical plan.)

xuechendi replied:
Cool, I'll work that. Will send out a PR later

BryanCutler replied:
Sounds like this is working towards similar things. @xuechendi the reason I brought up the intermediate conversion to bytes is that it's an expensive step and not necessary if you are just converting  `ArrowRecordBatch` <-> `ColumnarBatch`. It's done in `ArrowConverters` specifically to send/read Arrow messages over a socket with Spark.

HyukjinKwon replied:
Oh yeah. I think we shouldn't do the conversion step.

xuechendi replied:
@BryanCutler, are you saying converting between `ArrowRecordBatch` <-> `ColumnarBatch` is expensive? As far as I understand the codes, using Arrow `VectorSchemaRoot` to load `ArrowRecordBatch` just doing `ArrowBuf` reference_count increment then construct as a Arrow List[ValueVector], and this is the same method `ArrowConverters.fromBatchIterator` uses.
It will be great if you can help me to understand your expected way, thanks!
I would like to use any existing method in ArrowConverters or add some to it, the thing is the method you previously mentioned I replied before [`fromBatchIterator`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala#L132-L136) is converting to RDD[InternalRow] not RDD[ColumnarBatch](This is why I didn't use it), but I can extract the shared part of the codes as a new function, and called by both `fromBatchIterator`(RDD[InternalRow]) and a new function, let's name it as `toColumnarBatchIterator`(RDD[ColumnarBatch])
Do you think it makes sense?

xuechendi replied:
@HyukjinKwon , what is the byte conversion step you are mentioning, is this for Iterator[InternalRow] to Iterator[ArrowRecordBatch]  or Iterator[ArrowRecordBatch] to Iterator[ColumnarBatch](From my understanding, this part is not doing heavy memory copy, as replied with some detail in last comment) ?

BryanCutler reviewed (commented):
@xuechendi thanks for the PR, I'd like to hear a little more about your use case especially how you will with batches serialized to byte arrays, and I'm also curious how much this has in common with RAPIDS use of columnar batches.

xuechendi commented:
> @xuechendi thanks for the PR, I'd like to hear a little more about your use case especially how you will with batches serialized to byte arrays, and I'm also curious how much this has in common with RAPIDS use of columnar batches.

my use case is that we intend to use Arrow ColumnarBatch as our ColumnarBatch, and we have some operators implemented based on Arrow ColumnarBatch. if this PR is acceptable, then spark will support Arrow as a third candidate for ColumnarBatch Solution(since it can convert Row To Arrow Based Columnar).

I think our use case will be very similar with how RAPIDS use of columnar, so this PR should work for both side?

xuechendi committed:
Add a parameter to specify arrow as RowToColumnarExec backend

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala:
```diff
@@ -71,61 +71,114 @@ private[sql] class ArrowBatchStreamWriter(
    * in a batch by setting maxRecordsPerBatch or use 0 to fully consume rowIter.
    */
-  private[sql] def toBatchIterator(
+  private[sql] def toArrowRecordBatchIterator(
```
@BryanCutler and @HyukjinKwon , based on our previous discussion, I moved all arrow related conversions to `ArrowConverters`, so now I added or modified previous c onversions as below 6 functions
* `toArrowRecordBatchIterartor`  :   `Iterator[InternalRow]` -> `Iterator[ArrowRecordBatch]`
* `fromArrowRecordBatchIterartor`  :   `Iterator[ArrowRecordBatch]` -> `Iterator[InternalRow]`
* 
* `toBatchIterator` `Iterator[InteralRow]` -> `Iterator[Array[Byte]]` (will internally call `toArrowRecordBatchIterartor` )
* `fromBatchIterator` `Iterator[Array[Byte]]`-> `Iterator[InteralRow]` (will internally call `fromArrowRecordBatchIterator`)
* 
* `toColumnarBatchIterator` `Iterator[ArrowRecordBatch]` -> `Iterator[ColumnarBatch]` arrow backend
* `fromColumnarBatchIterator` `Iterator[ColumnarBatch]` -> `Iterator[ArrowRecordBatch]`

xuechendi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -458,6 +461,24 @@ case class RowToColumnarExec(child: SparkPlan) extends RowToColumnarTransition {
+          timeZoneId,
+          context)
+      }
+    }
```
@BryanCutler and @HyukjinKwon , for the implementation of RowToColumnarExec, I now use `toArrowRecordBatchIterator` + `toColumnarBatchIterator` to do the conversion, so there will be no serialization and deserialization.

xuechendi commented on sql/core/src/main/scala/org/apache/spark/sql/execution/Columnar.scala:
```diff
@@ -426,7 +428,8 @@ trait RowToColumnarTransition extends UnaryExecNode
  * would only be to reduce code.
  */
-case class RowToColumnarExec(child: SparkPlan) extends RowToColumnarTransition {
+case class RowToColumnarExec(child: SparkPlan, useArrow: Boolean = false)
```
@HyukjinKwon , after I added 4 extra conversion functions in `ArrowConverters`, I felt those can also be used by mapInArrow function, so I only added a parameter here to force specify Arrow as Columnar Backend

xuechendi commented on sql/core/src/test/scala/org/apache/spark/sql/execution/arrow/ArrowConvertersSuite.scala:
```diff
@@ -1393,6 +1393,55 @@ class ArrowConvertersSuite extends SharedSparkSession {
     assert(count == inputRows.length)
   }
 
+  test("convert from ArrowRecordBatch to ColumnarBatch") {
```
@BryanCutler and @HyukjinKwon , new UT is added in `ArrowConvertersSuite.scala`

xuechendi committed:
Add UT for ColumnarBatch to ArrowRecordBatch

Signed-off-by: Chendi Xue <chendi.xue@intel.com>

xuechendi commented:
@BryanCutler and @HyukjinKwon, Please help me take a review, I updated this PR per our discussion last week
Here is my last update
I moved all arrow related conversions to ArrowConverters, so now I added or modified previous c onversions as below 6 functions

toArrowRecordBatchIterartor : Iterator[InternalRow] -> Iterator[ArrowRecordBatch]
fromArrowRecordBatchIterartor : Iterator[ArrowRecordBatch] -> Iterator[InternalRow]
toBatchIterator Iterator[InteralRow] -> Iterator[Array[Byte]] (will internally call toArrowRecordBatchIterartor )
fromBatchIterator Iterator[Array[Byte]]-> Iterator[InteralRow] (will internally call fromArrowRecordBatchIterator)
toColumnarBatchIterator Iterator[ArrowRecordBatch] -> Iterator[ColumnarBatch] arrow backend
fromColumnarBatchIterator Iterator[ColumnarBatch] -> Iterator[ArrowRecordBatch]
